### ISSUE TITLE
eth_getFilterLogs returns an empty array

### ISSUE THREAD

**Issue by mpolci** [ConsenSys-archive/ganache-cli-archive#145]
It seems that this function doesn't return the past logs

Here a trace from the truffle console:

```
truffle(default)> var TXID='0x5a882f8a21443edb76868669ebaf49dc2990e414e8fc0c0d3acbbf409ad5f2ae'

truffle(default)> web3.eth.getTransactionReceipt(TXID).logs
[ { logIndex: 0,
    transactionIndex: 0,
    transactionHash: '0x5a882f8a21443edb76868669ebaf49dc2990e414e8fc0c0d3acbbf409ad5f2ae',
    blockHash: '0xe13a35edd48b8a2f5da82aa7e7dbd2b11fb0d1134f7c0252e1b4e33f6d4e0e8c',
    blockNumber: 9,
    address: '0xb6e2b8f8de940170d6389efc584529ecc1966c6b',
    data: '0x000000000000000000000000618fcd1ab22a91911613e622fc03f5ace2bf595d',
    topics: 
     [ '0xbb766b38574b4b06dacd1d1436133856d5f955e912882638c674397fa5a21ae1',      
       '0x000000000000000000000000ae0a1daf53184e02911f8a669fda449e20b471a4' ],
    type: 'mined' } ]
```

You can see the log fingerprint **'0xbb766b38574b4b06dacd1d1436133856d5f955e912882638c674397fa5a21ae1'**, it corresponds to:

```
truffle(default)> web3.sha3('NewProject(address,address)')
'0xbb766b38574b4b06dacd1d1436133856d5f955e912882638c674397fa5a21ae1'   

truffle(default)> FundingHub.abi[8]                                      
{ anonymous: false,
  inputs: 
   [ { indexed: true, name: 'owner', type: 'address' },
     { indexed: false, name: 'project', type: 'address' } ],
  name: 'NewProject',
  type: 'event' }

```

But this log is not returned by the `eth_getFilterLogs` rpc call:

```
truffle(default)> var contract_address = web3.eth.getTransaction(TXID).to

truffle(default)> var MyContract = web3.eth.contract(FundingHub.abi)

truffle(default)> var myContractInstance = MyContract.at(contract_address)

truffle(default)> var filter = myContractInstance.NewProject().get((err,log) => console.log('error:',err,' - log:', log))
undefined
error: null  - log: []      
```

The returned log array shouldn't be empty.

**Comment by tcoulter** [comment]
First, great use of the Truffle console. :) 

Second, the current spec for getFilterLogs (e.g., eth_newFilter) specifies different behavior than some Ethereum clients. The TestRPC follows this spec.

More specifically, check out the [default parameters for eth_newFilter](https://github.com/ethereum/wiki/wiki/JSON-RPC#eth_newfilter). If no `fromBlock` is specified, as is the case with your example above, then the fromBlock of "latest" is used. What this means is "Create me a filter that gives me logs from the current block and beyond." In your example above, its not clear whether or not the transaction you're referring to was made within the very last block. If it wasn't, it wouldn't show up in the logs.

To be very explicit and ensure you get the logs you expect, specify a fromBlock of 0, or something before or equal to block number 9 (in this case). You should receive the log you expect.

As a general rule, you should always create a filter (i.e., watch for an event, in web3 parlance) **before** triggering the event with a transaction. In your case, you've done the opposite.

I'm going to close this ticket for housekeeping purposes. However, if you're sure the transaction you're referring to happened in the very last block (i.e., double check with eth_blockNumber), and the log still isn't being shown, then it's likely a bug.

Cheers!

**Comment by mpolci** [comment]
Thank you for your quick answer. 

I repeated the test and I checked the block number, the block is the latest, I also tryed to pass the fromBlock parameter with no success:

```
truffle(default)> web3.eth.getTransactionReceipt(TXID).blockNumber
5

truffle(default)> web3.eth.blockNumber
5


truffle(default)> var filter = myContractInstance.NewProject({}, {fromBlock:0}).get((err,log) => console.log('error:',err,' - log:', log))
undefined
error: null  - log: []


truffle(default)> var filter = myContractInstance.NewProject(null, {fromBlock:4}).get((err,log) => console.log('error:',err,' - log:', log))
undefined
error: null  - log: []

truffle(default)> var filter = myContractInstance.NewProject(null, {fromBlock:'earliest'}).get((err,log) => console.log('error:',err,' - log:', log))
undefined
error: null  - log: []

truffle(default)> var filter = myContractInstance.NewProject(null, {fromBlock:'earliest', toBlock:'latest'}).get((err,log) => console.log('error:',err,' - log:', log))
undefined
error: null  - log: []
```

Just for information, the same code using a geth node works.

**Comment by tcoulter** [comment]
Can you tell me which version of the TestRPC you're using? It should say at the top when you run the testrpc.

To make things easier from a bug reporting standpoint, can you write a short snippet of Javascript code that reproduces the error, perhaps using the Truffle default project (or any contracts, whatever's easiest). That would make it much easier for me to reproduce and understand the full context of what's happening.

**Comment by mpolci** [comment]
The testrpc version is v2.2.1 (I just updated). To reproduce the error you can use this project:
https://github.com/mpolci/b9lab-final-project

The error is generated in the 3rd migration script, simply run the migration and see 
the message to discover the transaction hash generating the event:

```
user@ubuntu:~/final-project$ truffle migrate --reset
Running migration: 1_initial_migration.js
  Deploying Migrations...
  Migrations: 0x57e2f4d56574f321202607ed176786bf1b3a36f7
Saving successful migration to network...
Saving artifacts...
Running migration: 2_deploy_contracts.js
  Deploying FundingHub...
  FundingHub: 0x947bb4fb1c0d64c9bfabc3a9c361eb1066209ee6
Saving successful migration to network...
Saving artifacts...
Running migration: 3_create_first_project.js
  Running step...
  Create project transaction: 0x66e2c036aa5a0aea2fb6715959b44490f9cc681dc5700ed50cc94f82d9dea892
Error encountered, bailing. Network state unknown. Review successful transactions manually.
Project not created
```

Look for the message "Create project transaction: xxxxx".

**Comment by tcoulter** [comment]
I did some digging, and this does appear to be erroring because you didn't specify a `fromBlock`. More importantly though, it occurs because you're creating the event watcher **after** triggering the event (it doesn't make sense to listen for something after it's happened, right?). 

To make your code work, change [line 32 of 3_create_first_project.js](https://github.com/mpolci/b9lab-final-project/blob/testrpc-bug/migrations/3_create_first_project.js#L32) to this: 

```
FundingHub.deployed().NewProject({}, {fromBlock: 0}).get((err, res) => {
```

If you look at the [contract events API for web3](https://github.com/ethereum/wiki/wiki/JavaScript-API#contract-events), which Truffle uses, you'll notice the first object passed to NewProject is a set of values (empty) that change the topics used to filter events. The second object passed contains are `fromBlock`, and if you change the line, you do indeed get the event you were looking for.

However, it's bad practice, especially on the live chain, to get all events from block 0, as a lot can happen. Instead, I recommend you create the event watcher somewhere near line 12 before deployment happens:

```
var event = FundingHub.deployed().NewProject();
```

And then when you need to get the event, in `getCreatedProject` just call `event.get()` like you did previously. If you do it this, you don't have to specify a `fromBlock` because `fromBlock` defaults to whatever the current block is at time of creation.

I agree this behavior is rather clunky. It's on my radar to make events easier to use, as web3's events are more confusing than anything else. Cheers! Hope this helps.

**Comment by mpolci** [comment]
Thank you for your advices. 
I don't know why in my previous test in the truffle console, specifing the fromBlock didn't work. Probably I made a mistake, sorry for this. Now it works.

I was inganned because the code worked with geth, but evidently geth has a strange behavior. I did a little hard to understand how to use events on web3 and I was wrong :)

Finally, I agree, it is not a good practice to get all past events from block 0. Probably in my case is not a problem because the contract is just created and there is only one event (I suppose it is indexed by the account address), but I'm going to change the code to be more clear.
Thanks a lot
